File: objc/Demo Objc/ConnectBLEViewController.m
//
//  ConnectBLEViewController.m
//  Demo Objc
//
//  Created by Tatiana Magdalena on 24/10/17.
//  Copyright © 2017 Eduardo Mello de Macedo | Stone. All rights reserved.
//

#import "ConnectBLEViewController.h"


@interface ConnectBLEViewController ()

@property (strong, nonatomic) IBOutlet UITableView *tableView;
@property (strong, nonatomic) IBOutlet UILabel *feedback;

@property (strong, nonatomic) UIView *overlayView;
@property (strong, nonatomic) UIActivityIndicatorView *activityIndicator;

@end

@implementation ConnectBLEViewController

NSMutableArray <STNPinpad *> *peripherals;
STNPinPadConnectionProvider *connection;

#pragma mark - Lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    
    self.navigationItem.title = @"Conexão com BLE";
    
    self.overlayView = [[UIView alloc] initWithFrame:[UIScreen mainScreen].bounds];
    self.overlayView.backgroundColor = [UIColor colorWithRed:0 green:0 blue:0 alpha:0.5];
    self.activityIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
    self.activityIndicator.center = self.overlayView.center;
    
    if ([STNValidationProvider validatePinpadConnection] == NO) {
        NSLog(@"Pinpad não conectado.");
        self.feedback.text = @"Desconectado";
    } else {
        NSLog(@"Pinpad conectado.");
        self.feedback.text = @"Conectado";
    }

    connection = [[STNPinPadConnectionProvider alloc] init];
    connection.delegate = self;
    [connection startCentral];
    
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Buttons actions

- (IBAction)startScanning:(id)sender
{
    [connection startScan];
}

- (IBAction)disconnectAllBLE:(id)sender
{
    NSArray<STNPinpad*>* connectedPinpads = [connection listConnectedPinpads];
    
    for (STNPinpad* pinpad in connectedPinpads) {
        [connection disconnectPinpad:pinpad];
    }
}

#pragma mark - Tableview

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return peripherals.count;
}


-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"peripheralCell" forIndexPath:indexPath];
    cell.textLabel.text = peripherals[indexPath.row].name;
    return cell;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [connection connectToPinpad:peripherals[indexPath.row]];
    [connection stopScan];
}

#pragma mark - Pinpad Connection Delegate

-(void)pinpadConnectionProvider:(STNPinPadConnectionProvider *)provider didStartScanning:(BOOL)success error:(NSError *)error
{
    NSLog(@"$Did start scanning: %@", success ? @"true" : @"false");
    [self setFeedbackMessage:@"Começou a escanear"];
}

-(void)pinpadConnectionProvider:(STNPinPadConnectionProvider *)provider didFindPinpad:(STNPinpad *)pinpad
{
    if(peripherals == nil) {
        peripherals = [[NSMutableArray alloc] init];
    }
    
    if(![peripherals containsObject:pinpad]) {
        NSLog(@"$Did find pinpad: %@", pinpad.name);
        [peripherals addObject:pinpad];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.tableView reloadData];
        });
    }
}

-(void)pinpadConnectionProvider:(STNPinPadConnectionProvider *)provider didConnectPinpad:(STNPinpad *)pinpad error:(NSError * _Nullable)error
{
    NSLog(@"$Did connect to pinpad: %@", pinpad.name);
    [self setFeedbackMessage:@"Conectou ao pinpad"];
    //[connection disconnectPinpad:pinpad];
    
    [connection selectPinpad:pinpad];
}

-(void)pinpadConnectionProvider:(STNPinPadConnectionProvider *)provider didDisconnectPinpad:(STNPinpad *)pinpad
{
    NSLog(@"$Did disconnect from pinpad: %@", pinpad.name);
    [self setFeedbackMessage:@"Desconectou do pinpad"];
}

-(void)pinpadConnectionProvider:(STNPinPadConnectionProvider *)provider didChangeCentralState:(CBManagerState)state
{
    NSString *stateString = @"";
    
    switch (state) {
        case 0: stateString = @"Unknown"; break;
        case 1: stateString = @"Resetting"; break;
        case 2: stateString = @"Unsupported"; break;
        case 3: stateString = @"Unauthorized"; break;
        case 4: stateString = @"Powered off"; break;
        case 5: stateString = @"Powered on"; break;
        default:
            break;
    }
    
    NSLog(@"$Did change central state: %@", stateString);
}

-(void)setFeedbackMessage:(NSString*)message
{
    dispatch_async(dispatch_get_main_queue(), ^{
        self.feedback.text = message;
    });
}

@end

File: swift/Demo Swift/ConnectBLEViewController.swift
//
//  ConnectBLEViewController.swift
//  Demo Swift
//
//  Created by Kennedy Noia | Stone on 31/01/18.
//  Copyright © 2018 Stone Pagamentos. All rights reserved.
//

import UIKit
import StoneSDK

class ConnectBLEViewController: UIViewController, STNPinPadConnectionDelegate, UITableViewDelegate, UITableViewDataSource {
    
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var feedback: UILabel!
    
    var loadingView: LoadingView!
    
    var peripherals = [STNPinpad]()
    var connection: STNPinPadConnectionProvider!
    
    // MARK: Lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.delegate = self
        tableView.dataSource = self
        
        //Set loading overlay view
        self.loadingView = LoadingView.init(frame: UIScreen.main.bounds)
        self.navigationController?.view.addSubview(self.loadingView)
        
        connection = STNPinPadConnectionProvider.init()
        connection.delegate = self
        connection.startCentral()
    }
    
    // MARK: Buttons actions
    @IBAction func startScanning(_ sender: Any) {
        connection.startScan()
    }
    
    @IBAction func disconnectAllBLE(_ sender: Any) {
        let connectedPinpads = connection.listConnectedPinpads()
        
        connectedPinpads.forEach { pinpad in
            connection.disconnectPinpad(pinpad)
        }
    }
    
    // MARK: Tableview
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return peripherals.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "peripheralCell", for: indexPath)
        cell.textLabel?.text = peripherals[indexPath.row].name
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        connection.connect(to: peripherals[indexPath.row])
        connection.stopScan()
        setFeedback(message: "Conectado com \(peripherals[indexPath.row].name)")
    }
    
    // MARK: Pinpad Connection Delegate
    func pinpadConnectionProvider(_ provider: STNPinPadConnectionProvider, didStartScanning success: Bool, error: Error?) {
        setFeedback(message: "Começou a escanear")
    }
    
    func pinpadConnectionProvider(_ provider: STNPinPadConnectionProvider, didFind pinpad: STNPinpad) {
        if peripherals.contains(pinpad) == false
        {
            peripherals.append(pinpad)
            DispatchQueue.main.async {
                self.tableView.reloadData()
            }
        }
    }
    
    func pinpadConnectionProvider(_ provider: STNPinPadConnectionProvider,
                                  didConnect pinpad: STNPinpad, error: Error?) {
        connection.select(pinpad)
        self.setFeedback(message: "Conectou a pinpad")
    }
    
    func pinpadConnectionProvider(_ provider: STNPinPadConnectionProvider,
                                  didDisconnectPinpad pinpad: STNPinpad) {
        setFeedback(message: "Desconectou de pinpad")
    }
    
    func pinpadConnectionProvider(_ provider: STNPinPadConnectionProvider,
                                  didChangeCentralState state: CBManagerState) {
        var stateString = ""
        
        
        
        switch state {
        case .unknown:
            stateString = "Unknown"
            break
        case .resetting:
            stateString = "Resetting"
            break
        case .unsupported:
            stateString = "Unsupported"
            break
        case .unauthorized:
            stateString = "Unauthorized"
            break
        case .poweredOff:
            stateString = "Powered off"
            break
        case .poweredOn:
            stateString = "Powered on"
            break
        }
        print("Central state", stateString)
    }
    
    func setFeedback(message: String) {
        DispatchQueue.main.async {
            self.feedback.text = message
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}
